File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cpp
class Solution {
public:
    int myAtoi(string s) {
        return myAtoiRecursive(s, 0, 1, 0, false);
    }
private:
    int myAtoiRecursive(string& s, int index, int sign, long long result, bool signDetermined)
    {
        if(index >= s.size())
        {
            return result*sign;
        }
        
        
        char currentChar = s[index];
        
        if(!signDetermined && currentChar == ' ')
        {
            return myAtoiRecursive(s, index+1, sign, result, signDetermined);
        }
        
        if(!signDetermined && (currentChar == '-' || currentChar == '+'))
        {
            sign = (currentChar == '-') ? -1 : 1;
            return myAtoiRecursive(s, index+1, sign, result, true);
        }
        
        if(isdigit(currentChar))
        {
            result = result*10 + (currentChar - '0');
            if(sign * result > INT_MAX)
            {
                return INT_MAX;
            }
            
            if(sign * result < INT_MIN)
            {
                return INT_MIN;
            }

            return myAtoiRecursive(s, index+1, sign, result, true);
        }
        
        return sign*result;
    }
};

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.cs
public class Solution {
    public int MyAtoi(string s) {
        if (string.IsNullOrEmpty(s))
            return 0;

        int i = 0;
        int n = s.Length;
        
        while (i < n && s[i] == ' ')
            i++;
        
        int sign = 1;
        if (i < n && (s[i] == '-' || s[i] == '+'))
        {
            sign = s[i] == '-' ? -1 : 1;
            i++;
        }
        long result = 0;
        while (i < n && char.IsDigit(s[i]))
        {
            result = result * 10 + (s[i] - '0');
            if (sign * result > int.MaxValue)
                return int.MaxValue;
            if (sign * result < int.MinValue)
                return int.MinValue;
            i++;
        }
        
        return (int)(sign * result);
    }
}